### ISSUE TITLE
Performing actions (such as lock, unlock etc) fails with a traceback

### ISSUE THREAD

**Issue by stellarhopper** [alandtse/tesla#774]
## Version of the custom_component

v3.19.3

## Configuration

Not sure what to attach here, but using the default config for everything.

```yaml

```

## Describe the bug

This is a fresh install of Home Assistant on the 'Yellow' board, and of this plugin. I'm able to sync data from tesla correctly, but performing any action such as locking/unlocking doors, setting charge limit etc doesn't seem to work. There is a traceback in the logs that is attached below.

## Debug log

```text

2023-11-12 17:10:09.296 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [548106961344] 
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 230, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 2035, in async_call
    response_data = await coro
                    ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2072, in _execute_service
    return await target(service_call)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 235, in handle_service
    return await service.entity_service_call(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 876, in entity_service_call
    response_data = await _handle_entity_call(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 948, in _handle_entity_call
    result = await task
             ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/lock/__init__.py", line 99, in _async_lock
    await entity.async_lock(**remove_entity_service_fields(service_call))
  File "/config/custom_components/tesla_custom/lock.py", line 36, in async_lock
    await self._car.lock()
  File "/usr/local/lib/python3.11/site-packages/teslajsonpy/car.py", line 854, in lock
    data = await self._send_command("LOCK")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/teslajsonpy/car.py", line 775, in _send_command
    raise ex
  File "/usr/local/lib/python3.11/site-packages/teslajsonpy/car.py", line 760, in _send_command
    data = await self._controller.api(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/teslajsonpy/controller.py", line 1311, in api
    response = await self.__post_with_retries(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/tenacity/_asyncio.py", line 88, in async_wrapped
    return await fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/tenacity/_asyncio.py", line 47, in __call__
    do = self.iter(retry_state=retry_state)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/tenacity/__init__.py", line 325, in iter
    raise retry_exc.reraise()
          ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/tenacity/__init__.py", line 158, in reraise
    raise self.last_attempt.result()
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.11/site-packages/tenacity/_asyncio.py", line 50, in __call__
    result = await fn(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/teslajsonpy/controller.py", line 1334, in __post_with_retries
    return await self.__connection.post(command, method=method, data=data, url=url)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/teslajsonpy/connection.py", line 165, in post
    return await self.__open(url, method=method, headers=self.head, data=data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/teslajsonpy/connection.py", line 216, in __open
    raise TeslaException(resp.status_code)
teslajsonpy.exceptions.TeslaException

```

**Comment by alandtse** [comment]
Please enable debugging and then provide updated logs of the call.

**Comment by stellarhopper** [comment]
FWIW The Tesla software version is 2023.38.6

**Comment by efoverskov** [comment]
Just adding to this - same issue for me - brand new MY FW 2023.32.200. When trying any action or calling an action (SCHEDULED_CHARGING) it fails with error
`[teslajsonpy.connection] 403: {"response":null,"error":"Tesla Vehicle Command Protocol required, please refer to the documentation here: https://developer.tesla.com/docs/fleet-api#2023-10-09-rest-api-vehicle-commands-endpoint-deprecation-warning","error_description":""}`
Could this be Tesla implementing this on new vehicles? Polling works - same API call worked on my old 2020 M3.

**Comment by joperezr** [comment]
I'm getting the same thing, worked fine this morning but started to fail this afternoon with:

``` 
2023-11-13 19:48:20.837 DEBUG (MainThread) [teslajsonpy.connection] 403: {"response":null,"error":"Tesla Vehicle Command Protocol required, please refer to the documentation here: https://developer.tesla.com/docs/fleet-api#2023-10-09-rest-api-vehicle-commands-endpoint-deprecation-warning","error_description":""}
```

**Comment by alandtse** [comment]
Yes, this is a known issue #742. Tesla is breaking the old API for newer vehicles. Hopefully someone who has such a vehicle will submit a PR.

**Comment by joperezr** [comment]
IIUC the new API is a Go SDK which AFAIK isn't trivially usable from Python, so I guess one would need to first write python bindings for that SDK, so that it can be used by this integration.

**Comment by KevinThibaut89** [comment]
> Yes, this is a known issue #742. Tesla is breaking the old API for newer vehicles. Hopefully someone who has such a vehicle will submit a PR.

Brand new Model Y here (got it last Friday). Same problem. 
Anything I could do to help?

**Comment by mmeyerc** [comment]
Got my new Tesla MYLR last week softwareversion 2023.32.200 i have same problem. All data comes into Home Assistant fine, but I can't send anything the other way

**Comment by joperezr** [comment]
So I think I found a potential solution to this. According to https://github.com/teslamotors/vehicle-command, the repo not only contains the new Go command SDK, but instead it also provides [an http proxy](https://github.com/teslamotors/vehicle-command#using-the-http-proxy) that can be setup which can receive calls using the old FLEET API, and can then internally use the go SDK in order to forward those commands appropriately so that they can start reaching the car again. The setup for this proxy is for sure not straight forward, and it involves generating private keys, and registering those with your car via a new tesla endpoint, but assuming we get over the setup hurdle, we should be able to continue to make the calls using the FLEET API and for those to work with new cars that require the new SDK. The overall plan would be basically this for folks that want to set this up:

1. Users interested in this, will have to setup a server (either a new raspberry pi or the same one running home assistant in case its running in docker) which is running the HTTP proxy and is properly configured following [these instructions](https://github.com/teslamotors/vehicle-command#distributing-your-public-key).
2. Then, users will go and configure the tesla home assistant integration in this repo by following the regular instructions in the README. One new requirement (feature request) would be that one new configuration setting is exposed by the integration, which is the url where the commands will be sent against. I only looked briefly at teslajsonpy APIs but it does look like they support passing in alternate urls for sending the commands. If this setting is provided to the integration, then the integration can just use that url when initializing the tesla controller, so that all calls are forwarded to it. Users will configure the url to be the address of the server that they have setup in step 1.
3. Everything should work now 😃 

This solution is of course not ideal because it will require a lot of setup from the user's side, but the main benefit is that it would require very minimal changes on this integration (just exposing the new setting and flowing it through) and will immediately add support to all of the cars that were built post-2021 (which of course will be quite a bit XD).

@alandtse thoughts? Does this sound reasonable?

**Comment by alandtse** [comment]
Happy to take a PR. If someone wants to create a HA Addon for the proxy, we can leverage that.

**Comment by daddyd0108** [comment]
is there any chance to get that to work?

**Comment by alandtse** [comment]
Unfortunately you'll probably have to wait for someone with the new car that requires this to submit a PR or workaround. I'm not planning on buying a new Tesla (and doubt anyone would buy one for me ;) ) and the old api will continue to work for me.

**Comment by stellarhopper** [comment]
Btw I don't think this is so much related to a new car rather than maybe
Tesla flipping a switch as to when the old API gets deprecated?

Mine is a 2021 Y.

**Comment by lotharbach** [comment]
> The setup for this proxy is for sure not straight forward, and it involves generating private keys, and registering those with your car via a new tesla endpoint, but assuming we get over the setup hurdle, we should be able to continue to make the calls using the FLEET API and for those to work with new cars that require the new SDK.

The big caveat here is that Tesla has to approve every single Fleet-API client. Which they likely won't do for individual vehicle owners. So all of this effort might end up only useful for a select few that by chance managed to gain Fleet-API access.

There are two solutions that "scale" and could allow every vehicle owner to continue to send commands:
  * Find out how the Tesla App is doing it, which I am trying and invite more collaborators in https://github.com/timdorr/tesla-api/discussions/769 - this would then send the command directly to Teslas servers but would also require a complicated setup procedure to register a key to the vehicle via Bluetooth.
  * One or multiple "trusted" Fleet-API "Partners" grant everybody access "through" them. As the TOS and common sense forbid sharing their credentials and keys directly, they would need to run the proxy on their infrastructure, which means you would need to authorize and trust them with more or less unlimited vehicle control.  So you would send a variation of "/api/1/vehicles/{id}/command/charge_start" to say Teslascope's or maybe TeslaFi's API, they turn it into a signed command with their private key, and send that to the Fleet API with their partner credentials.

**Comment by gcamp** [comment]
A third solution that scale (but with different tradeoffs) : 
- Use the BLE integration from https://github.com/teslamotors/vehicle-command to have local polling of your Tesla. Benefit : local polling. Drawback : only works when you Tesla is at Home and that you have BLE range where it's parked.

**Comment by darek-margas** [comment]
> Unfortunately you'll probably have to wait for someone with the new car that requires this to submit a PR or workaround. I'm not planning on buying a new Tesla (and doubt anyone would buy one for me ;) ) and the old api will continue to work for me.

It won't take long for yours. Update from ChargeHQ below:

UPDATE 2023-11-24

Tesla have announced that the old API will stop working on vehicles as follows:

    November 2023 -newly delivered vehicles will only support the new API
    Nov - Dec 2023 - the old API will stop working on existing vehicles that have not used the old API in the preceding 30 days
    January 2024 - the old API will stop working on all vehicles

**Comment by sjvc** [comment]
What about using the Tessie API? I know it's not free, but they are using the official Tesla API. We could set up the Tessie token, and make API calls to our Teslas through Tessie API.

**Comment by juicejuice** [comment]
> only works when your Tesla is at Home and that you have BLE range where it's parked.

I guess someone could also cook up what is essentially a WAN-to-BLE gateway that you can install in your car to allow access outside of the home? I wonder if that gains you much over just using the new API.

**Comment by lgkahn** [comment]
i maintain a program in hubitat that used to work for controlling the car.. now that seems also broken for newer cars as outlined above.. i see teslafi has an alexa integration.. hopefully they will make this continue to work.. and maybe even do their own local api.. i fear that this will be the only type of viable solutions in the future a big aggregator like them that handles the backend with tesla and you go through them to do commands... pain int he ass.

**Comment by danielnitsche** [comment]
I’ve got some experience in golang but not much with home assistant, and I have one of the new affected teslas. 

If anyone wants to work together on this let me know - I’ve successfully got Tesla control working on a mac, but haven’t attempted with a raspberry pi (which is what I use for HA). I’d be looking at the BLE solution only, and mostly interested in changing options rather than full control.

**Comment by motley-git** [comment]
> I’ve got some experience in golang but not much with home assistant, and I have one of the new affected teslas.
> 
> If anyone wants to work together on this let me know - I’ve successfully got Tesla control working on a mac, but haven’t attempted with a raspberry pi (which is what I use for HA). I’d be looking at the BLE solution only, and mostly interested in changing options rather than full control.

Hi, I have an affected 2021 M3P + a working HA in docker (Intel platform) currently using https://github.com/alandtse/tesla (all options are read-only).  I'll be looking at the BLE integration @gcamp posting as I have time to see if it does what I need to send some simple commands.  Start/stop charging commands will be my priority since I will be working on ab automation and/or script in HA that allows stop/start charging based upon Tesla charge level,  current solar production, home battery storage charge level etc.  I'd be interested in working with others to test/learn (on this or other synergistic projects) and perhaps even write some code once I have a decent foundation and better understand the codebase, goals etc.

**Comment by lgkahn** [comment]
i am just ready to beta test a new version of tesla control i have implemented for the hubitat home automation.. this is using the teslafi interface so you have to sign up with them.
currently it works if you are stll on the OLD tesla api version, but i have talked to teslafi and they will be modifying their implementation to work with the new protocoll/api that is rumored to take full control at the end of january.

due to the fact that tesla will be monitizing this and you need to be a company to effectively use the new api you will need a third party integrator like teslafi..

thus my implementation i hope will be transparent when they are done and the new api kicks fully in

**Comment by darek-margas** [comment]
Recent core added Tessie integration and it works.

**Comment by dcegelis** [comment]
> Recent core added Tessie integration and it works.

But requires a tessie subscription

**Comment by NyLan-1** [comment]
Tesla has shut down the API we've been relying on...

@alandtse Is there any news, developments, or even a glimmer of hope regarding this integration working again in the future? Any updates or insights you can provide would be greatly appreciated.

**Comment by JasSmiths** [comment]
Just as an FYI this now appears to be affecting Powerwall as well as vehicles. 

Everything is read only.

**Comment by lgkahn** [comment]
I.implemented a new home automation integration in hubitat via www.Tessie.com if anyone is interested.

This to replace my prior integration that went direct to tesla.

**Comment by haggertd** [comment]
> I.implemented a new home automation integration in hubitat via [www.Tessie.com](http://www.Tessie.com) if anyone is interested.
> 
> This to replace my prior integration that went direct to tesla.

Isn't Tessie a paid service though?

**Comment by mr-p666** [comment]
Unfortunately yes.

**Comment by dennypage** [comment]
Yes, it's a paid service. However, the way Tesla has structured things you are going to require some form of service provider going forward regardless. The direct to the end customer model is ending.

**Comment by danielnitsche** [comment]
> Yes, it's a paid service. However, the way Tesla has structured things you are going to require some form of service provider going forward regardless. The direct to the end customer model is ending.

BLE is still an option locally

**Comment by jaymunro** [comment]
> The direct to the end customer model is ending.

Yes. It seems from this (https://developer.tesla.com/docs) that they now require some legal business credentials before granting access.

> Click the ‘Submit Request’ button below to request app access. After signing in, get started and provide your legal business details, app name, description and purpose of usage. Upon submitting your details, we will review your request and email you with status updates.

**Comment by jaymunro** [comment]
> The direct to the end customer model is ending.

Yes. It seems from this (https://developer.tesla.com/docs) that they now require some legal business credentials before granting access.

> Click the ‘Submit Request’ button below to request app access. After signing in, get started and provide your legal business details, app name, description and purpose of usage. Upon submitting your details, we will review your request and email you with status updates.

From this I interpret it as they are specifically excluding the open source model for web access. Perhaps too likely to be abused by crackers??

As such, it seems BLE is the way forward (https://github.com/teslamotors/vehicle-command) to at least allow home charging automation and other local configuration. Personally I don't use HA when away from home with the car to control the car, I use the car to do that as I'm usually in it.

In that repo they have the following:
> Among the included tools is an HTTP proxy server that converts REST API calls to the new vehicle-command protocol.

Perhaps this can be leveraged to bridge the gap with little effort to have a working system again?

**Comment by jaymunro** [comment]
The just released [Tessie HA integration](https://www.home-assistant.io/integrations/tessie/) is looking tempting, and at $50 a year it could just replace my TeslaFi subscription for the same cost.
> The Tessie integration was introduced in Home Assistant 2024.1, and it's used by [175](https://analytics.home-assistant.io/integrations) active installations.
Will have to do some more research.

**Comment by lgkahn** [comment]
as i said also here
https://community.hubitat.com/t/control-and-query-your-tesla-vehicles-via-tessie-tesla-connect-replacement/
but think you need to be a member of the hubitat forums to see it

![Screenshot 2024-01-26 160508](https://github.com/alandtse/tesla/assets/236665/299fe7f0-d583-4ecd-a3e5-dd008fa485bc)

**Comment by kimme1024** [comment]
I wouldn’t mind buying a Tessie subscription if that would mean the solar charging script would keep on working for the next 5 years. That saves me a lot more.

However if free would remain an option that’s obviously preferable as I don’t have any use for the Tessie service itself.

**Comment by jaymunro** [comment]
I'm sold. I've opened an account and will be stopping my TeslaFi sub.
You can get a free month with this referral https://share.tessie.com/74PWQ3xrGGb

**Comment by kimme1024** [comment]
> > I wouldn’t mind buying a Tessie subscription if that would mean the solar charging script would keep on working for the next 5 years. That saves me a lot more.
> 
> I'm sold. I've opened an account and will be stopping my TeslaFi sub. You can get a free month with this referral https://share.tessie.com/74PWQ3xrGGb

I’ll wait for a couple of days to see what’s going to be done here. As for me, $200 for a lifetime sub still seems like a lot of money to regain access to the force update button. (It’s the only thing that stops my usecase from working)

**Comment by baylanger** [comment]
> I wouldn’t mind buying a Tessie subscription if that would mean the solar charging script would keep on working for the next 5 years. That saves me a lot more.

We should try to get a group discount :D 

Somehow I wonder if in near future T won't shift all traffic to their new Fleet API and turn off their old API just to reduce the cost associated to the "extra" traffic.

**Comment by Frenchie90-alt** [comment]
Hi,

I have the new Tessie integration to Home Assistant and from what I can tell, it only has Wake (This only wakes the car if it's asleep) it does not a force to update to get latest info.

But from what I can see it updates everything as soon as it's updated from the tesla end.

**Comment by jaymunro** [comment]
> Somehow I wonder if in near future T won't shift all traffic to their new Fleet API and turn off their old API just to reduce the cost associated to the "extra" traffic.

See https://github.com/alandtse/tesla/issues/774#issuecomment-1829773375
>November 2023 -newly delivered vehicles will only support the new API
Nov - Dec 2023 - the old API will stop working on existing vehicles that have not used the old API in the preceding 30 days
January 2024 - the old API will stop working on all vehicles

**Comment by baylanger** [comment]
> > Somehow I wonder if in near future T won't shift all traffic to their new Fleet API and turn off their old API just to reduce the cost associated to the "extra" traffic.
> 
> See [#774 (comment)](https://github.com/alandtse/tesla/issues/774#issuecomment-1829773375)
>
I thought the announcement was only for the "vehicle commands" to shit to the new Fleet API. Sounds like the telemetry calls will shutdown soon, perhaps as soon as Jan 31? Long life batteries included but not free API :/

**Comment by darek-margas** [comment]
> > I wouldn’t mind buying a Tessie subscription if that would mean the solar charging script would keep on working for the next 5 years. That saves me a lot more.
> 
> I'm sold. I've opened an account and will be stopping my TeslaFi sub. You can get a free month with this referral https://share.tessie.com/74PWQ3xrGGb

On Reddit there is permanent discount code which he published early days.

**Comment by darek-margas** [comment]
> I’ll wait for a couple of days to see what’s going to be done here. As for me, $200 for a lifetime sub still seems like a lot of money to regain access to the force update button. (It’s the only thing that stops my usecase from working)
There won't be update button as it does noit make sense. This is man-in-the middle, HA-Tessie is just current and updating from car to Tessie would be the other side task, not in HA. Yet somehow it is current so maybe it has more clever pulling?

**Comment by darek-margas** [comment]
> Just as an FYI this now appears to be affecting Powerwall as well as vehicles.
> 
> Everything is read only.

Isn't that the PW has local API able to do everything?

**Comment by thierryvt** [comment]
> > I wouldn’t mind buying a Tessie subscription if that would mean the solar charging script would keep on working for the next 5 years. That saves me a lot more.
> 
> We should try to get a group discount :D
> 
> Somehow I wonder if in near future T won't shift all traffic to their new Fleet API and turn off their old API just to reduce the cost associated to the "extra" traffic.

Has already happened for me, tried adding this integration to HA yesterday, which worked for all of 30 minutes. Then it just stopped working, sensors all started reading "unknown" and commands would just given an error. Debug logging shows this:

412: {"response":null,"error":"Endpoint is only available on fleetapi....

**Comment by mpepping** [comment]
```
2024-01-27 11:24:43.024 DEBUG (MainThread) [teslajsonpy.connection] get: https://owner-api.teslamotors.com/api/1/vehicles {}
2024-01-27 11:24:43.225 DEBUG (MainThread) [teslajsonpy.connection] 412: {"response":null,"error":"Endpoint is only available on fleetapi. Visit https://developer.tesla.com/docs for more info","error_description":""}
2024-01-27 11:24:43.226 DEBUG (MainThread) [custom_components.tesla_custom] Finished fetching tesla_custom data in 15.205 seconds (success: False)
```

**Comment by JasSmiths** [comment]
> > Just as an FYI this now appears to be affecting Powerwall as well as vehicles.
> > Everything is read only.
> 
> Isn't that the PW has local API able to do everything?

You can do the majority via the local API but not everything unfortunately

**Comment by gkwok1** [comment]
I made a few changes to teslajsonpy and got it working for my Pre-2021 Model S/X. My fork can be found here.

https://github.com/gkwok1/teslajsonpy/archive/refs/heads/dev.zip

Copy the teslajsonpy folder inside the zip file to /config on HA and restart.

I did not create a PR as this is not a solution for everyone.

Note from Tesla:
January 2024 - All vehicles* will require Tesla Vehicle Command protocol. The REST API will be fully deprecated
*Fleet accounts are excluded from these changes until further notice. Pre-2021 Model S/X are excluded from these changes.

**Comment by ve6rah** [comment]
> I did not create a PR as this is not a solution for everyone.
> 
> Note from Tesla: January 2024 - All vehicles* will require Tesla Vehicle Command protocol. The REST API will be fully deprecated *Fleet accounts are excluded from these changes until further notice. Pre-2021 Model S/X are excluded from these changes.

It seems like it should be a solution for all pre-2021 cars though, and there are MANY of us out there. It seems a PR is in order!

**Comment by gkwok1** [comment]
@ve6rah there are talks about other fixes as well and I am in no way skilled enough to understand all of the possible solution, I just want to share my code for what worked for my situation. There are probably better ways to do what I did to fix it as well.

**Comment by skipishere** [comment]
@gkwok1 feel free to raise a PR, most people in open source would be happy to have people add to (or try) fixes. The best part of a PR is when you get feedback about what does (or doesn't) work.

**Comment by ve6rah** [comment]
The issue here is that us pre-2021 S/X owners are being drowned out by the 3/Y and newer S/X owners. From what I can gather from Tesla, what works for them WILL NOT work for us and vice-versa. If we want pre-2021 cars to work at all, I think you can ignore the "other" fixes as they aren't relevant, and we should focus on getting yours implemented. The fix for newer cars and 3/Y cars is going to be a lot more in depth, and waiting for their fix is unlikely to help us anyway.

**Comment by gkwok1** [comment]
@ve6rah @skipishere feel free to take my code and create a PR for it, I just threw something together so it works for me.

**Comment by kimme1024** [comment]
> I made a few changes to teslajsonpy and got it working for my Pre-2021 Model S/X. My fork can be found here.
> 
> https://github.com/gkwok1/teslajsonpy/archive/refs/heads/dev.zip
> 
> Copy the teslajsonpy folder inside the zip file to /config on HA and restart.
> 
> I did not create a PR as this is not a solution for everyone.
> 
> Note from Tesla: January 2024 - All vehicles* will require Tesla Vehicle Command protocol. The REST API will be fully deprecated *Fleet accounts are excluded from these changes until further notice. Pre-2021 Model S/X are excluded from these changes.

Nice! I'd be happy to test this as I'm having a 2018 S 75D. I'll try to find the time next week to install it.
As you're saying that pre 2021 S/X are excluded from these changes, would this mean that for these cars everything will keep working for the forseeable future using your changes?

**Comment by gkwok1** [comment]
@kimme1024 looks like @iancg already submitted a PR to fix the issues for pre-2021 S/X wiht teslajsonpy.

As with all Tesla things, it will work until Elon changes his mind.

**Comment by ikerhg88** [comment]
Is there any solution for this issue?

**Comment by darek-margas** [comment]
> Is there any solution for this issue?

Tessie integration for now.

**Comment by thierryvt** [comment]
> > The direct to the end customer model is ending.
> 
> Yes. It seems from this (https://developer.tesla.com/docs) that they now require some legal business credentials before granting access.
> 
> > Click the ‘Submit Request’ button below to request app access. After signing in, get started and provide your legal business details, app name, description and purpose of usage. Upon submitting your details, we will review your request and email you with status updates.
> 
> From this I interpret it as they are specifically excluding the open source model for web access. Perhaps too likely to be abused by crackers??
> 
> As such, it seems BLE is the way forward (https://github.com/teslamotors/vehicle-command) to at least allow home charging automation and other local configuration. Personally I don't use HA when away from home with the car to control the car, I use the car to do that as I'm usually in it.
> 
> In that repo they have the following:
> 
> > Among the included tools is an HTTP proxy server that converts REST API calls to the new vehicle-command protocol.
> 
> Perhaps this can be leveraged to bridge the gap with little effort to have a working system again?

I've been looking into the HTTP proxy situation and managed to set up an account and register an application.
Still working on setting things up though because it is A LOT more involved and I wouldn't recommend it if you're not technically inclined and familiar with Oauth flows.

Not sure how well the HTTP proxy matches with the old REST API implementation though. I guess worst case I'll have to fork this repo and do some extra coding work to adapt to it.

I don't have that much time to invest into this so don't expect a solution by tonight but I'll post an update if I have one.

BTW according to this thread: https://github.com/teslamotors/vehicle-command/issues/148 the business credentials should be optional. So the fact they're required is either a bug or that guy doesn't know what he's talking about lol.

**Comment by iancg** [comment]
> BTW according to this thread: https://github.com/teslamotors/vehicle-command/issues/148 the business credentials should be optional. So the fact they're required is either a bug or that guy doesn't know what he's talking about lol.

I tried it a couple of weeks ago at which point if you are setting up a new account it did seem to require business credentials, but if you are an owner, you can add the dev features to your existing Tesla account by logging in to [https://developer.tesla.com/](https://developer.tesla.com/) with your existing account.  I have a legacy pre-21 MX, so in the end, this doesn't seem to apply to me so I couldn't test any further.

**Comment by thierryvt** [comment]
> > BTW according to this thread: [teslamotors/vehicle-command#148](https://github.com/teslamotors/vehicle-command/issues/148) the business credentials should be optional. So the fact they're required is either a bug or that guy doesn't know what he's talking about lol.
> 
> I tried it a couple of weeks ago at which point if you are setting up a new account it did seem to require business credentials, but if you are an owner, you can add the dev features to your existing Tesla account by logging in to https://developer.tesla.com/ with your existing account. I have a legacy pre-21 MX, so in the end, this doesn't seem to apply to me so I couldn't test any further.

I did log in with my existing account but it still asked me for business credentials and wouldn't let me continue without entering them.
I'm a freelancer though so I actually have business credentials.

**Comment by llamafilm** [comment]
I built an add-on that runs the official HTTP Proxy using your own personal Tesla auth and key.  This is my first time building an add-on so feedback is welcome.  I handle the initial authorization flow in a temporary Flask app, in conjunction with the Nginx SSL add-on.  Perhaps this integration could be modified to use the proxy.

https://github.com/llamafilm/tesla-http-proxy-addon

**Comment by thierryvt** [comment]
> I built an add-on that runs the official HTTP Proxy using your own personal Tesla auth and key. This is my first time building an add-on so feedback is welcome. I handle the initial authorization flow in a temporary Flask app, in conjunction with the Nginx SSL add-on. Perhaps this integration could be modified to use the proxy.
> 
> https://github.com/llamafilm/tesla-http-proxy-addon

That's not running the http proxy, it just handles the oauth flow, still beat me to it lol.
Code-wise I see nothing immediately wrong but I also have no experience in creating HA plugins.

There's only 2 things I'm curious about:
1. seems like the public key is only hosted as long as the temporary Flask app is running, the documentation states it should always be available. Not sure if this could have side effects (because idk if it's only used for authentication or also for the refresh token or even the signing of commands).
2. you're using a self-signed ssl cert which is also not recommended "in production" as it could cause issues even if it's just you.

**Comment by llamafilm** [comment]
@thierryvt thanks for your feedback.  The public key is hosted forever using Nginx with valid certificate; that part does not go through Flask.  For example mine is https://tesla.llamafilm.com/.well-known/appspecific/com.tesla.3p.public-key.pem
I used a self-signed cert on the proxy because it's _only_ accessible locally, by other Docker containers on the same host (i.e this integration).

**Comment by llamafilm** [comment]
@alandtse I'm still trying to work it all out, but I think teslajsonpy does not need any changes for my proxy to work.  I can make it trust the local cert by changing util.py in this addon:
```py
    import ssl
    context = ssl.create_default_context()
    SSL_CONTEXT = context.load_verify_locations(cafile='/share/tesla/selfsigned.pem')
```

It looks like this integration tries to exchange the refresh token for an access token, but the HTTP Proxy does not support that.  So we need to split it into 2 different URLs.  First hit `https://auth.tesla.com/oauth2/v3/token` to exchange the refresh token for an access token.  Then use that access token to send commands to the local proxy.  Does that sound reasonable?

**Comment by skipishere** [comment]
Update for those that had issues with the refresh button not working: the latest update fixes that (just tested it myself - M3 2023)

**Comment by jcsp101** [comment]
> Update for those that had issues with the refresh button not working: the latest update fixes that (just tested it myself - M3 2023)

The latest update for the integration? Or what?

**Comment by skipishere** [comment]
Have updated the comment to make it clearer I meant the integration

**Comment by Fonsohome** [comment]
Hello, I don't know if it can help but the teslemetry integration does work although it is only for the weather at the moment.  The integration uses an access token and a virtual key which is granted with the official Tesla application.

**Comment by llamafilm** [comment]
Interesting.  It looks like Teslemetry has implemented pretty much the same solution I did, but they charge a fee and make it easy to use for folks who don't have their own domain name.

**Comment by darek-margas** [comment]
> I prefer not to trust strangers with full remote access to my vehicle. By giving them your API token and installing their public key in your vehicle, you are giving Teslemetry full access to your vehicle — just like Teslafi and other services have done in the past.

That's exactly why I avoided them and also why I am happy Tesla make these changes so that access given can be adjusted to what is needed only.

**Comment by thierryvt** [comment]
@tugakufi teslajsonpy fix has just been released, waiting on llamafilm to finalize changes on this integration. If all goes well it should be ready this weekend. You can keep an eye on this PR: https://github.com/alandtse/tesla/pull/853 

Or you could pull my fork of this repo (https://github.com/thierryvt/tesla) if you really can't wait.

Do however note that the new API set-up is quite a bit more involved than the current one.

**Comment by darek-margas** [comment]
I spent time today to get proxy up in docker container. I don't use supervised HA just stockpile of containers sandwiched by docker-compose. Now it seems up. Added key to the car and wondering - is this integration ready to switch over? What would be best method to re-authenticate and while doing it, to change to use proxy? Other than reinstalling it.
BTW - proxy listens locally on 443 - is it required? I mean would this integration be able to hit proxy on different port?

**Comment by llamafilm** [comment]
This integration is now working fine for me and a few others. So I guess this issue can be closed. Just update to the latest version and follow the instructions. One issue still remaining is there’s a bug in the reauth config flow. So to reathenticate, you need to delete the car from HA and add it back. 

@darek-margas you can change the port if you want.

**Comment by kmil4** [comment]
But it's not working fine?... I have had zero issues for over a year. Now I can't send commands on the latest versions but I can read sensor data.

**Comment by jcsp101** [comment]
Same here, not working for me either. I can see sensor status but get fail to call service error on any command

**Comment by llamafilm** [comment]
Did you read the README? Specifically the section about Fleet API Proxy. This was a major breaking API change from Tesla.

**Comment by kmil4** [comment]
Yes, especially the part about the log error to look for, which is not the log error that I'm getting. And I'm not in a new car. People keep insisting that all issues raised tie back to he fleet API.

And if they do, the README docs should be updated to say to reflect that.

**Comment by skipishere** [comment]
> Yes, especially the part about the log error to look for, which is not the log error that I'm getting. And I'm not in a new car. People keep insisting that all issues raised tie back to he fleet API.
> 
> And if they do, the README docs should be updated to say to reflect that.

If it is auth issue and the fleet API does work for you, you can make a pull request to update the docs.

It would be great if we could easily signpost in error messages that people are hitting the API issue, but given the amount of work that was initially done just to get the fleet API working we're lucky to even have what we've got.

Also you say you don't have a new car but from what Tesla have initially said, at some point most model of cars will have to migrate over, some of us have been fortunate enough it hasn't happened (as yet, model 3 2023 version here) but it is totally down to Tesla and how they go about forcing users from one system to the other.

**Comment by llamafilm** [comment]
FWIW, my 2021 model 3 requires the new API. It seems like this is being rolled out incrementally.

**Comment by brettjenkins** [comment]
> Yes, especially the part about the log error to look for, which is not the log error that I'm getting. And I'm not in a new car. People keep insisting that all issues raised tie back to he fleet API.
> 
> And if they do, the README docs should be updated to say to reflect that.

This, I know it's not the authentication because some commands (like flashing lights and turning on climate) work fine. Surely they'd all fail if I haven't set up the virtual key and proxy correctly? So climate and flashing lights work fine, but stuff like setting heated seats and steering wheel fails. Any idea?

**Comment by skipishere** [comment]
@brettjenkins your error message is different `Error decoding response into json` please open a new issue

**Comment by cyleo** [comment]
I just finished setting up the http-proxy, and the integration seems to be working, but now I get the same issue: 
`2024-05-06 21:41:53.265 DEBUG (MainThread) [teslajsonpy.connection] post: https://owner-api.teslamotors.com/api/1/vehicles/LRW3E7FSXRC070226/command/flash_lights {'on': True}
2024-05-06 21:41:53.320 DEBUG (MainThread) [teslajsonpy.connection] 403: {"response":null,"error":"Tesla Vehicle Command Protocol required, please refer to the documentation here: https://developer.tesla.com/docs/fleet-api#2023-10-09-rest-api-vehicle-commands-endpoint-deprecation-warning","error_description":""}
2024-05-06 21:41:55.513 DEBUG (MainThread) [custom_components.tesla_custom] Running controller.update()
2024-05-06 21:41:55.514 DEBUG (MainThread) [teslajsonpy.controller] Get vehicles. Force: False Time: 61 Interval 60
2024-05-06 21:41:55.514 DEBUG (MainThread) [teslajsonpy.connection] Token expiration in 7:38:01
2024-05-06 21:41:55.514 DEBUG (MainThread) [teslajsonpy.connection] get: https://owner-api.teslamotors.com/api/1/products {}`

**Comment by llamafilm** [comment]
@cyleo that log message means you are not using the proxy.  I suggest removing the integration and adding it again, making sure to tick the box for proxy and enter your Client ID.

**Comment by cyleo** [comment]
> @cyleo that log message means you are not using the proxy. I suggest removing the integration and adding it again, making sure to tick the box for proxy and enter your Client ID.

Thanks! I had to use the fleet api button in Auth for Tesla which I overlooked.

**Comment by kiler129** [comment]
It appears that the restriction is rolled into old cars as well. As exposing HA instance to the internet is a bit of a risky move imo.

I'm reading through multiple documentation resources that have conflicting information. If I'm understanding it correctly:
- Fleet API is the new API that isn't implemented in most software at this point
- [The official API proxy](https://github.com/teslamotors/vehicle-command) allows the use for the legacy/owner API by translating the calls
- The Fleet API requires mutual OAuth and this is handled by the HTTP proxy
- To run the proxy one needs to essentially register app with tesla, which requires HTTP server always serving the public key for the integration. The HTTP server doesn't need to respond to any calls by itself (so it can be hosted on like S3?). The official HTTP proxy is not related to that part of the setup.
- There is no official Docker container from Tesla for HTTP proxy, but there [are simple ones available](https://github.com/rileymd88/vehicle-command/blob/main/Dockerfile)
- I can write some documentation on that if needed, to give some more details beyond "install another 3rd party addon and forward your port to HA"


Given that we're dealing with cars and controlling them here, I think the security should be treated quite seriously ;)
